### ISSUE TITLE
broken compatibility with Silicon-testing-framework

### ISSUE THREAD

**Issue by aiace9** [libAtoms/QUIP#174]
From the point of view of a PhD student involved in the creation and study of new silicon potentials I would like to compare my results with what already obtained by your group. To do so the easiest way is to run the [silicon testing framework](https://github.com/libAtoms/silicon-testing-framework) provided with the paper. Since the framework strongly relays on quippy, here I am opening this issue.

# Failed command:
Now I am trying to run the following command

`python3 run-model-test.py SW bulk_diamond`

As is in _my version_ of the [testing framework](https://github.com/aiace9/silicon-testing-framework) under the test-runner folder. 

Within my testing framework I am referring to commit 982a453 that is a modification on top of the current master commit ec91b17fb. 
Strictly speaking, at the moment,  my branch differs form the master for few things:
- it adapts all the prints and the imports to py3
- it moves the `quippy import Potential` to `quippy.potential import Potential`

# Test with your docker container
 sha: 60b496ca1a
(small bug: as I am writing your doc main page under pre-compiled images  points [here](https://github.com/libAtoms/QUIP/blob/public/docker/README.md), file that is no more available)
In this case I get this error

> Traceback (most recent call last):
>   File "run-model-test.py", line 96, in <module>
>     import test  # import and run the current test
>   File "../../tests/bulk_diamond/test.py", line 25, in <module>
>     (c11, c12, c44, E_vs_V) = lattice_cubic.do_lattice(bulk, elastic=True)
>   File "../../share/lattice_cubic.py", line 31, in do_lattice
>     bulk = relax_atoms_cell(bulk, tol=tol, traj_file=None, symmetrize=True)
>   File "../../share/utilities.py", line 158, in relax_atoms_cell
>     opt.run(tol, max_steps, smax=stol)
>   File "/usr/local/lib/python3.6/site-packages/ase/optimize/precon/lbfgs.py", line 376, in run
>     return Optimizer.run(self, fmax, steps)
>   File "/usr/local/lib/python3.6/site-packages/ase/optimize/optimize.py", line 246, in run
>     return Dynamics.run(self)
>   File "/usr/local/lib/python3.6/site-packages/ase/optimize/optimize.py", line 159, in run
>     for converged in Dynamics.irun(self):
>   File "/usr/local/lib/python3.6/site-packages/ase/optimize/optimize.py", line 131, in irun
>     self.log()
>   File "/usr/local/lib/python3.6/site-packages/ase/optimize/precon/lbfgs.py", line 392, in log
>     e = self.atoms.get_potential_energy()
>   File "/usr/local/lib/python3.6/site-packages/ase/constraints.py", line 1991, in get_potential_energy
>     force_consistent=force_consistent)
>   File "/usr/local/lib/python3.6/site-packages/ase/atoms.py", line 669, in get_potential_energy
>     self, force_consistent=force_consistent)
>   File "/usr/local/lib/python3.6/site-packages/ase/calculators/calculator.py", line 626, in get_potential_energy
>     'not provided by {0} calculator'.format(name))
> ase.calculators.calculator.PropertyNotImplementedError: Force consistent/free energy ("free_energy") not provided by SymmetrizedCalculator calculator

That crashed the code at the very beginning during the optimization step.

# Test with my docker container
I also created a container by myself to have more control over what is happening. My container is very minimal and forces an installation of quippy with only python 3 so it requires a small patch on the current commit [^1].
You can replicate my environment by using the following files:

- [Dockerfile](https://pastebin.com/tuewU2H2)
- [patch_py3.diff](https://pastebin.com/a3cbhuWW)
- [config.txt](https://pastebin.com/swriC6kU)
- GAP version 1572365433

In this case the first error is caused by an IO instruction in lattice_cubic.py:34.

> Traceback (most recent call last):
>   File "./run-model-test.py", line 96, in <module>
>     import test  # import and run the current test
>   File "../../tests/bulk_diamond/test.py", line 25, in <module>
>     (c11, c12, c44, E_vs_V) = lattice_cubic.do_lattice(bulk, elastic=True)
>   File "../../share/lattice_cubic.py", line 34, in do_lattice
>     ase.io.write(sys.stdout, bulk, format='extxyz')
>   File "/usr/lib/python3/dist-packages/ase/io/formats.py", line 352, in write
>     _write(filename, fd, format, io, images, parallel=parallel, **kwargs)
>   File "/usr/lib/python3/dist-packages/ase/io/formats.py", line 381, in _write
>     io.write(fd, images, **kwargs)
>   File "/usr/lib/python3/dist-packages/ase/io/extxyz.py", line 624, in write_xyz
>     per_frame_results)
>   File "/usr/lib/python3/dist-packages/ase/io/extxyz.py", line 517, in output_column_format
>     dtype = np.dtype(dtypes)
> ValueError: field 'forces0' occurs more than once

Commenting that line, that should not be crucial at the moment, I get this much more beefy error

> Traceback (most recent call last):
>   File "./run-model-test.py", line 96, in <module>
>     import test  # import and run the current test
>   File "../../tests/bulk_diamond/test.py", line 25, in <module>
>     (c11, c12, c44, E_vs_V) = lattice_cubic.do_lattice(bulk, elastic=True)
>   File "../../share/lattice_cubic.py", line 55, in do_lattice
>     scaled_bulk = relax_atoms_cell(scaled_bulk, tol=tol, traj_file=None, constant_volume=True, method='cg_n', symmetrize=True)
>   File "../../share/utilities.py", line 117, in relax_atoms_cell
>     from quippy.potential import Minim, fzeros
> ImportError: cannot import name 'Minim'

I found the Minim routine in the Fortran code and, as far as I understood, this routine should be exposed somehow in the potential file that is(?) generated by f90wrap. For some reason this is not happening and something similar may be happening with fzeros. 
Can you confirm the errors and check if there are simple viable solutions?

Thanks, 
Ruggero.

[^1]:  the commit I am referring to is a03cda33 and it is not enforced by the docker file

**Comment by jameskermode** [comment]
Thanks for reporting. Quippy has now moved on to Python 3, but this framework uses the old Python2.7 version, which you can obtain from this branch: https://github.com/libAtoms/QUIP/tree/archived_py2

There is not currently a legacy version of the Docker container, but we could look to set one up if you are unable to reproduce our results using this branch.

**Comment by aiace9** [comment]
The misalignment between the python versions of the 2 codes was the first thing I noticed and I promptly tired your py2 tag that should correspond to the archived_py2 branch.
The test was done with the following configuration:

- [Dockerfile](https://pastebin.com/wV7GC0y7)
- GAP version 1572365433
- [patch](https://pastebin.com/uPH2M5zr) required by GAP to compile

and resulted in what looked like a memory leak since when loading quippy module in to python RAM usage went up to 5gb. (At that point I stopped it)

At the end of the test I decided that, since it is very likely that I'll have to add some code on top of the testing-framework and the 1 of January is now very close it was meaningful to move on and concentrate on solving the problems in a python3 environment rather than a python2. This is also why I haven't included the results of this test in the issue before.

If you already have all the code versions that should work together it could be worth considering to create a container. Otherwise, if there aren't many changes, we can look in to updating the framework to the last quippy version.

Thanks for your help.
